Add 'indexof', 'startswith', 'endswith' and 'length' string operations

derekwolpert opened the pull request:
Additional string operations to assist in creating more complex conditionals in Noop Logic.

**indexof**:
`{ "indexof": [ "bar", "foobar" ] } // 3`
`{ "indexof": [ "foobar", "bar" ] } // -1`

**startswith**:
`{ "startswith": [ "foo", "foobar" ] } // true`
`{ "startswith": [ "bar", "foobar" ] } // false`

**endswith**:
`{ "endswith": [ "foo", "foobar" ] } // false`
`{ "endswith": [ "bar", "foobar" ] } // true`

**length**:
`{ "length": [ "foobar" ] } // 6`
`{ "length": [ "" ] } // 0`

mwkaufman commented:
Looks good to me. It looks like we're using camelCase more than snake_case so it would seem reasonable to be consistent with one or the other in outward facing Noop properties, but I'll leave it to @arbitrarytech 's input there.

derekwolpert commented:
@mwkaufman JSON Logic uses snake case instead of camelcase, and Noop Logic inherits rules from JSON Logic that use snake case (e.g. `missing_some`).

derekwolpert commented:
@mwkaufman @arbitrarytech 

Quick overview on changes in the latest commit:

- Changed `index_of`, `starts_with`, and `ends_with` to `indexof`, `startswith`, and `endswith`. This was done to match @arbitrarytech 's name choice for `useragent`. Whether the operation names would be snake_case, camelCase or something else altogether is still a decision that needs to be settled on.

- Added support to pass in an Array into the `cidr` operation. If the given `ip` matches any of the CIDRs in the array the operation returns `true`.
  - e.g. `{"cidr": [["10.0.0.0/32", "10.0.0.1/32", "10.0.0.2/32"], "10.0.0.1"]} // true`

- Added support for optional `prop` param with `useragent` and `qs` operation. When a `prop` is corresponding property of the resulting object is returned. Basically a shortcut for accessing a `prop` of the resulting object.  When no `prop` is supplied the operations execute the same as before.
  - existing syntax: `{"prop": ["os.family", {"useragent": {"var": "user-agent"}}]} // "Mac OS X"`
  - prop shortcut: `{"useragent": [{"var": "user-agent"}, ""os.family""]} // "Mac OS X"`

- Converted `prop` to class method to be used in multiple operations (included the `prop` operation).

- Updated code comments explain functionality of operations

arbitrarytech commented on index.js:
```diff
@@ -7,26 +7,80 @@ const dot = require('dot-object')
+    // Returns a boolean
+    logic.add_operation('startswith', (a, b) => {
+      if (!b || typeof b.indexOf === 'undefined') return false
+      return b.indexOf(a) === 0
```
Why not use the String.startsWith() method here?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith

derekwolpert replied:
@arbitrarytech I was initially thinking about converting removing ES5/6-isms from the code since the original JSON Logic code was written in roughly "raw" javascript. Didn't end up completely that 100%. I can swap it to use `startsWith` and `endsWith`, but functionality wise it'd be the same either way.

derekwolpert replied:
Obviously in a browser, as long as the code is transpiled it wouldn't matter either way.

arbitrarytech commented on index.js:
```diff
@@ -7,26 +7,80 @@ const dot = require('dot-object')
+    // Returns a boolean
+    logic.add_operation('endswith', (a, b) => {
+      if (!b || typeof b.indexOf === 'undefined') return false
+      return b.substr(b.length - a.length) === a
```
Same as above with String.endsWith()

arbitrarytech reviewed (changes requested):
Nice work @derekwolpert! I really like the additional “prop” arguments. Very minor changes in review.
